sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -55,99 +55,114 @@ object ConstantFolding extends Rule[LogicalPlan] {
+        val newAnd = if ((replacedNewLeft fastEquals left) && (replacedNewRight fastEquals right)) {
+          a
         } else {
-          if (newLeft.isDefined || newRight.isDefined) {
-            Some(And(newLeft.getOrElse(a.left), newRight.getOrElse(a.right)))
-          } else {
-            None
-          }
+          And(replacedNewLeft, replacedNewRight)
         }
-        (newSelf, equalityPredicates)
-      case o: Or =>
+        (newAnd, equalityPredicatesLeft ++= equalityPredicatesRight)
+      case o @ Or(left, right) =>
         // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newLeft, _) = traverse(o.left, replaceChildren = true)
-        val (newRight, _) = traverse(o.right, replaceChildren = true)
-        val newSelf = if (newLeft.isDefined || newRight.isDefined) {
-          Some(Or(left = newLeft.getOrElse(o.left), right = newRight.getOrElse((o.right))))
+        val (newLeft, _) = traverse(left)
+        val (newRight, _) = traverse(right)
+        val newOr = if ((newLeft fastEquals left) && (newRight fastEquals right)) {
+          o
         } else {
-          None
+          Or(newLeft, newRight)
         }
-        (newSelf, Seq.empty)
-      case n: Not =>
+
+        (newOr, Map.empty)
+      case n @ Not(child) =>
         // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newChild, _) = traverse(n.child, replaceChildren = true)
-        (newChild.map(Not), Seq.empty)
-      case _ => (None, Seq.empty)
+        val (newChild, _) = traverse(child)
+        val newNot = if (newChild fastEquals child) {
+          n
+        } else {
+          Not(newChild)
+        }
+        (newNot, Map.empty)
+      case _ => (expression, Map.empty)
     }
 
-  private def replaceConstants(condition: Expression, equalityPredicates: EqualityPredicates)
-    : Expression = {
-    val constantsMap = AttributeMap(equalityPredicates.map(_._1))
-    val predicates = equalityPredicates.map(_._2).toSet
-    def replaceConstants0(expression: Expression) = expression transform {
-      case a: AttributeReference => constantsMap.getOrElse(a, a)
-    }
-    condition transform {
-      case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
-      case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
+  private def replaceConstants(
+      expression: Expression,
+      constants: Map[Expression, Literal]) =
+    expression transformUp {
+      case e if constants.contains(e.canonicalized) => constants(e.canonicalized)
     }
+}
+
+/**
+ * Substitutes expressions which can be statically reduced by constraints.
+ * eg.
+ * {{{
+ *   SELECT * FROM table WHERE i <= 5 AND i = 5        => ... WHERE i = 5
+ *   SELECT * FROM table WHERE i < j AND ... AND i > j => ... WHERE false
+ * }}}
+ */
+object FilterReduction extends Rule[LogicalPlan] with ConstraintHelper {
```

Comment:
Please focus on `FilterReduction` here independently.

Reply:
Ok, I've moved all new filter reduction related logic to `FilterReduction`.